File: app/services/triggers.py
from __future__ import annotations

import re
from collections.abc import Iterable

from aiogram.types import Message, MessageEntity

# Default trigger pattern for backwards compatibility
# Will be overridden by persona configuration or BOT_TRIGGER_PATTERNS setting
_DEFAULT_TRIGGER_PATTERN = re.compile(
    r"\b(?:гр[яи]г[аоуеєіїюяьґ]*|gr[yi]ag\w*)\b", re.IGNORECASE | re.UNICODE
)

# Global trigger patterns (set by initialize_triggers function)
_TRIGGER_PATTERNS: list[re.Pattern[str]] = [_DEFAULT_TRIGGER_PATTERN]


def initialize_triggers(patterns: list[str] | None = None) -> None:
    """Initialize trigger patterns from configuration.

    Args:
        patterns: List of regex pattern strings. If None, uses default pattern.
    """
    global _TRIGGER_PATTERNS

    if patterns:
        _TRIGGER_PATTERNS = [
            re.compile(pattern, re.IGNORECASE | re.UNICODE) for pattern in patterns
        ]
    else:
        _TRIGGER_PATTERNS = [_DEFAULT_TRIGGER_PATTERN]


def _contains_keyword(text: str | None) -> bool:
    if not text:
        return False
    # Check all trigger patterns
    return any(pattern.search(text) for pattern in _TRIGGER_PATTERNS)


def _matches_mention(
    text: str | None,
    entities: Iterable[MessageEntity] | None,
    username: str | None,
    bot_id: int | None,
) -> bool:
    if not text or not entities:
        return False
    target = username.lstrip("@").lower() if username else ""
    for entity in entities:
        if entity.type == "mention":
            mention = text[entity.offset : entity.offset + entity.length]
            if mention.lstrip("@").lower() == target:
                return True
        if entity.type == "text_mention" and entity.user:
            if bot_id is not None and entity.user.id == bot_id:
                return True
            if (
                target
                and entity.user.username
                and entity.user.username.lower() == target
            ):
                return True
    return False


def addressed_to_bot(
    message: Message,
    bot_username: str,
    bot_id: int | None = None,
    chat_id: int | None = None,
) -> bool:
    """Return True if the incoming message is directed to the bot.

    Args:
        message: The incoming message
        bot_username: Bot's username
        bot_id: Bot's user ID
        chat_id: Chat ID (positive for personal chats, negative for groups).
                 If positive (personal chat), keyword trigger is not required.
    """

    username = (bot_username or "").lstrip("@").lower()

    if username:
        if message.reply_to_message and message.reply_to_message.from_user:
            reply_user = message.reply_to_message.from_user
            reply_username = (reply_user.username or "").lower()
            if reply_username == username or (
                bot_id is not None and reply_user.id == bot_id
            ):
                return True

        if _matches_mention(message.text, message.entities, username, bot_id):
            return True
        if _matches_mention(
            message.caption, message.caption_entities, username, bot_id
        ):
            return True

    if _contains_keyword(message.text) or _contains_keyword(message.caption):
        return True

    # In personal chats (chat_id > 0), respond without keyword trigger
    if chat_id is not None and chat_id > 0:
        return True

    return False

File: cpp/src/services/triggers.cpp
#include "gryag/services/triggers.hpp"

#include <algorithm>
#include <cctype>

namespace gryag::services {

namespace {

// Default trigger pattern for Ukrainian and English variations
// Matches: гряг, грiг, gryag, griag, etc. with various endings
constexpr const char* kDefaultPattern = R"(\b(?:гр[яи]г[аоуеєіїюяьґ]*|gr[yi]ag\w*)\b)";

}  // namespace

TriggerDetector::TriggerDetector(const std::vector<std::string>& patterns) {
    if (patterns.empty()) {
        // Use default pattern
        trigger_patterns_.emplace_back(
            kDefaultPattern,
            std::regex::icase | std::regex::ECMAScript
        );
    } else {
        // Compile user-provided patterns
        for (const auto& pattern : patterns) {
            trigger_patterns_.emplace_back(
                pattern,
                std::regex::icase | std::regex::ECMAScript
            );
        }
    }
}

bool TriggerDetector::addressed_to_bot(
    const telegram::Message& message,
    const std::string& bot_username,
    std::int64_t bot_id
) const {
    const auto username = normalize_username(bot_username);

    // Check if message is a reply to the bot
    if (!username.empty() && message.reply_to_user.has_value()) {
        const auto& reply_user = message.reply_to_user.value();
        const auto reply_username = normalize_username(reply_user.username);

        if (reply_username == username || reply_user.id == bot_id) {
            return true;
        }
    }

    // Check for @mentions in message text
    if (!message.text.empty() && !message.entities.empty()) {
        if (matches_mention(message.text, message.entities, username, bot_id)) {
            return true;
        }
    }

    // Check for @mentions in media caption
    if (!message.caption.empty() && !message.caption_entities.empty()) {
        if (matches_mention(message.caption, message.caption_entities, username, bot_id)) {
            return true;
        }
    }

    // Check for keyword triggers in text
    if (!message.text.empty() && contains_keyword(message.text)) {
        return true;
    }

    // Check for keyword triggers in caption
    if (!message.caption.empty() && contains_keyword(message.caption)) {
        return true;
    }

    // In private chats (chat_id > 0), always respond
    if (message.chat.id > 0) {
        return true;
    }

    return false;
}

bool TriggerDetector::contains_keyword(const std::string& text) const {
    if (text.empty()) {
        return false;
    }

    // Check if any trigger pattern matches
    for (const auto& pattern : trigger_patterns_) {
        if (std::regex_search(text, pattern)) {
            return true;
        }
    }

    return false;
}

bool TriggerDetector::matches_mention(
    const std::string& text,
    const std::vector<telegram::MessageEntity>& entities,
    const std::string& username,
    std::int64_t bot_id
) const {
    if (text.empty() || entities.empty() || username.empty()) {
        return false;
    }

    for (const auto& entity : entities) {
        // Check text @mention (e.g., "@botname")
        if (entity.type == "mention") {
            // Extract the mentioned username from text
            if (entity.offset + entity.length <= text.size()) {
                const auto mention = text.substr(entity.offset, entity.length);
                const auto normalized_mention = normalize_username(mention);

                if (normalized_mention == username) {
                    return true;
                }
            }
        }

        // Check text_mention (clickable mention with user object)
        if (entity.type == "text_mention" && entity.user.has_value()) {
            const auto& mentioned_user = entity.user.value();

            // Check by user ID
            if (mentioned_user.id == bot_id) {
                return true;
            }

            // Check by username
            const auto mentioned_username = normalize_username(mentioned_user.username);
            if (!mentioned_username.empty() && mentioned_username == username) {
                return true;
            }
        }
    }

    return false;
}

std::string TriggerDetector::normalize_username(const std::string& username) {
    if (username.empty()) {
        return "";
    }

    // Remove @ prefix if present
    auto result = username;
    if (result[0] == '@') {
        result = result.substr(1);
    }

    // Convert to lowercase
    std::transform(result.begin(), result.end(), result.begin(),
                   [](unsigned char c) { return std::tolower(c); });

    return result;
}

}  // namespace gryag::services
